### PR TITLE
Allow using langchain4j module

### PR DESCRIPTION
**Pull request by anunnakian**
I add langchain4j support by using langchain4j-spring-starter dependency (version: 0.35.0)

It fixes https://github.com/jhipster/jhipster-lite/issues/9298

### CODE REVIEW

**Comment by anunnakian** [comment]
@pascalgrimaud @jdubois is that PR ok for you

File: src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# LangChain4J
+
+How to use it :
```
Comment (qmonmert) [inline src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache:3]:
```suggestion
How to use it:
```

File: src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  private static final JHipsterSource SOURCE = from("server/springboot/langchain4j");
+  private static final GroupId LANGCHAIN4J_GROUP_ID = groupId("dev.langchain4j");
+  private static final VersionSlug LANGCHAIN4J_VERSION_SLUG = versionSlug("langchain4j");
+
+  private static final String PROPERTIES = "properties";
+
+  public JHipsterModule buildModule(JHipsterModuleProperties properties) {
+    Assert.notNull(PROPERTIES, properties);
+
+    //@formatter:off
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java:23]: the formatter is off here, and it's intended. But you need to manually format the code below if possible

File: src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/infrastructure/primary/LangChain4JModuleConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+@Configuration
+class LangChain4JModuleConfiguration {
+
+  @Bean
+  JHipsterModuleResource langChain4JModule(LangChain4JApplicationService langChain4J) {
+    return JHipsterModuleResource.builder()
+      .slug(LANGCHAIN4J)
+      .propertiesDefinition(JHipsterModulePropertiesDefinition.builder().addProjectBaseName().addIndentation().build())
+      .apiDoc("LangChain4J", "Add LangChain4J Spring dependency")
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/infrastructure/primary/LangChain4JModuleConfiguration.java:21]:
```suggestion
      .apiDoc("LangChain4J", "Add LangChain4j")
```

File: src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# LangChain4J
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache:1]:
```suggestion
# LangChain4j
```

File: src/test/features/server/springboot/langchain4j.feature [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Feature: LangChain4J module
```
Comment (pascalgrimaud) [inline src/test/features/server/springboot/langchain4j.feature:1]:
```suggestion
Feature: LangChain4j module
```

File: src/test/features/server/springboot/langchain4j.feature [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+  Scenario: Should add Spring Boot LangChain4J Starter
```
Comment (pascalgrimaud) [inline src/test/features/server/springboot/langchain4j.feature:3]:
```suggestion
  Scenario: Should add Spring Boot LangChain4j starter
```

**Comment by pascalgrimaud** [comment]
In another PR, I would add a new module `sample-langchain4j-controller`, with something like:

```java
import dev.langchain4j.model.chat.ChatLanguageModel;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
class ChatController {

  private final ChatLanguageModel chatLanguageModel;

  public ChatController(ChatLanguageModel chatLanguageModel) {
    this.chatLanguageModel = chatLanguageModel;
  }

  @GetMapping("/chat")
  public String model(@RequestParam(value = "message", defaultValue = "Hello") String message) {
    return chatLanguageModel.generate(message);
  }
}
```

So it will help the user to test it (similar to the documentation):

image

File: src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        artifactId("langchain4j-spring-boot-starter"),
+        LANGCHAIN4J_VERSION_SLUG)
+      .addDependency(LANGCHAIN4J_GROUP_ID,
+        artifactId("langchain4j-open-ai-spring-boot-starter"),
+        LANGCHAIN4J_VERSION_SLUG)
+      .and()
+      .files()
+      .and()
+      .springMainProperties()
+      .set(propertyKey("langchain4j.open-ai.chat-model.api-key"), propertyValue("${OPENAI_API_KEY}"))
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java:37]: you can complete this part by adding a comment: `# You can temporarily use 'demo' key, which is provided for free for demonstration purposes`

File: src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache (hunk 4 above)
Comment (jdubois) [reply src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache:1]: Oh I do that mistake all the time!!!!!

**Comment by jdubois** [comment]
Super happy to see this! As I'm mostly using Azure OpenAI and GitHub Models, I'll need to have a look at adding those on top of this PR.

File: src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache (hunk 4 above)
Comment (anunnakian) [reply src/main/resources/generator/server/springboot/langchain4j/langchain4j.md.mustache:1]: CamelCase syndrome 😄

**Comment by anunnakian** [comment]
> In another PR, I would add a new module `sample-langchain4j-controller`, with something like:
> 
> ```java
> import dev.langchain4j.model.chat.ChatLanguageModel;
> import org.springframework.web.bind.annotation.GetMapping;
> import org.springframework.web.bind.annotation.RequestMapping;
> import org.springframework.web.bind.annotation.RequestParam;
> import org.springframework.web.bind.annotation.RestController;
> 
> @RestController
> @RequestMapping("/api")
> class ChatController {
> 
>   private final ChatLanguageModel chatLanguageModel;
> 
>   public ChatController(ChatLanguageModel chatLanguageModel) {
>     this.chatLanguageModel = chatLanguageModel;
>   }
> 
>   @GetMapping("/chat")
>   public String model(@RequestParam(value = "message", defaultValue = "Hello") String message) {
>     return chatLanguageModel.generate(message);
>   }
> }
> ```

So you'll have to link _langChain4j_ module to _spring-boot-mvc-empty_ because you'll need _rest_ dependency

File: src/test/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        """
+      )
+      .and()
+      .hasFile("src/main/resources/config/application.yml")
+      .containing(
+        """
+        langchain4j:
+          open-ai:
+            chat-model:
+              # # You can temporarily use 'demo' key, which is provided for free for demonstration purposes
```
Comment (pascalgrimaud) [inline src/test/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactoryTest.java:59]:
```suggestion
              # You can temporarily use 'demo' key, which is provided for free for demonstration purposes
```

File: src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import tech.jhipster.lite.module.domain.javabuild.ArtifactId;
+import tech.jhipster.lite.module.domain.javabuild.GroupId;
+import tech.jhipster.lite.module.domain.javabuild.VersionSlug;
+import tech.jhipster.lite.module.domain.properties.JHipsterModuleProperties;
+import tech.jhipster.lite.shared.error.domain.Assert;
+
+public class LangChain4JModuleFactory {
+
+  private static final String API_KEY_DEMO_COMMENT =
+    "# You can temporarily use 'demo' key, which is provided " + "for free for demonstration purposes";
```
Comment (pascalgrimaud) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/langchain4j/domain/LangChain4JModuleFactory.java:16]:
```suggestion
    "You can temporarily use 'demo' key, which is provided " + "for free for demonstration purposes";
```

**Comment by pascalgrimaud** [comment]
> > In another PR, I would add a new module `sample-langchain4j-controller`, with something like:
> > ```java
> > import dev.langchain4j.model.chat.ChatLanguageModel;
> > import org.springframework.web.bind.annotation.GetMapping;
> > import org.springframework.web.bind.annotation.RequestMapping;
> > import org.springframework.web.bind.annotation.RequestParam;
> > import org.springframework.web.bind.annotation.RestController;
> > 
> > @RestController
> > @RequestMapping("/api")
> > class ChatController {
> > 
> >   private final ChatLanguageModel chatLanguageModel;
> > 
> >   public ChatController(ChatLanguageModel chatLanguageModel) {
> >     this.chatLanguageModel = chatLanguageModel;
> >   }
> > 
> >   @GetMapping("/chat")
> >   public String model(@RequestParam(value = "message", defaultValue = "Hello") String message) {
> >     return chatLanguageModel.generate(message);
> >   }
> > }
> > ```
> 
> So you'll have to make _langChain4j_ module dependent on _spring-boot-mvc-empty_ because you'll need _rest_ dependency

exactly @anunnakian :-)

**Comment by pascalgrimaud** [comment]
Thanks a lot for your contribution @anunnakian !

**Comment by anunnakian** [comment]
https://opencollective.com/generator-jhipster/expenses/227229

**Comment by pascalgrimaud** [comment]
@anunnakian : approved

**Pull request merged**